File: python/leetcodehard/longest_valid_parentheses.py
"""
https://leetcode.com/problems/longest-valid-parentheses/

status - completed
"""

from base_logger.logging_event import create_logger


class Solution:

    log = create_logger(__name__)

    def longestValidParentheses(self, s):
        max_length = 0
        stack = [-1]

        for index, char in enumerate(s):
            if char == "(":
                stack.append(index)
            else:
                stack.pop()
                if not stack:
                    stack.append(index)
                else:
                    max_length = max(max_length, index - stack[-1])

        return max_length


if __name__ == "__main__":
    test_input_1 = "(()"
    output = Solution().longestValidParentheses(test_input_1)
    Solution().log.info("Output is %s", output)

    test_input_2 = ")()())"
    output = Solution().longestValidParentheses(test_input_2)
    Solution().log.info("Output is %s", output)

File: typescript/leetcodehard/longestValidParentheses.ts
/*
https://leetcode.com/problems/longest-valid-parentheses/

status - completed
*/

import { logp } from "../utils/logger";

function longestValidParentheses(s: string): number {
  const stack: number[] = [-1];
  let maxLen = 0;

  for (let i = 0; i < s.length; i++) {
    if (s[i] === "(") {
      stack.push(i);
    } else {
      stack.pop();
      if (stack.length === 0) {
        stack.push(i);
      } else {
        maxLen = Math.max(maxLen, i - stack[stack.length - 1]);
      }
    }
  }

  return maxLen;
}

logp(`input "(()" expected 2 actual ${longestValidParentheses("(()")}`);
logp(`input ")()())" expected 4 actual ${longestValidParentheses(")()())")}`);
logp(`input "" expected 0 actual ${longestValidParentheses("")}`);
